clippy_lints/src/swap.rs

```diff
@@ -70,9 +71,107 @@ impl<'tcx> LateLintPass<'tcx> for Swap {
     fn check_block(&mut self, cx: &LateContext<'tcx>, block: &'tcx Block<'_>) {
         check_manual_swap(cx, block);
         check_suspicious_swap(cx, block);
+        check_xor_swap(cx, block);
     }
 }
 
+fn subpart_stack<'a, 'hir>(mut e: &'a Expr<'hir>) -> Vec<&'a Expr<'hir>> {
+    let mut result = vec![];
+    loop {
+        result.push(e);
+        e = if let ExprKind::Index(ep, _) = e.kind {
+            ep
+        } else if let ExprKind::Field(ep, _) = e.kind {
+            ep
+        } else {
+            break;
+        }
+    }
+    result.reverse();
+    result
+}
+
+fn is_swappable(cx: &LateContext<'_>, e1: &Expr<'_>, e2: &Expr<'_>) -> bool {
+    let s1 = subpart_stack(e1);
+    let s2 = subpart_stack(e2);
+    if !eq_expr_value(cx, s1[0], s2[0]) {
+        return true;
+    }
+    for (x1, x2) in s1.iter().zip(s2.iter()) {
+        if !eq_expr_value(cx, x1, x2) {
+            match (&x1.kind, &x2.kind) {
+                (ExprKind::Field(_, _), ExprKind::Field(_, _)) => return true,
+                _ => return false,
+            }
+        }
+    }
+    false
+}
+
+fn generate_swap_warning(cx: &LateContext<'_>, e1: &Expr<'_>, e2: &Expr<'_>, span: Span, is_xor_based: bool) {
+    let mut applicability = Applicability::MachineApplicable;
+
+    if !is_swappable(cx, e1, e2) {
+        if let ExprKind::Index(lhs1, idx1) = e1.kind {
+            if let ExprKind::Index(lhs2, idx2) = e2.kind {
+                if eq_expr_value(cx, lhs1, lhs2) {
```

Comment:
Another `eq_expr_value` should not be necessary at this point. You may need to change `is_swappable` to having three possible outcomes (again). Suggestion: `Option<Swappable>`, `enum Swappable { MemSwap, SliceSwap }`. I know that is effectively the same as `Slice` but I think much clearer.

Reply:
Okay, that's fair. You could rename `is_swappable` to `can_mut_borrow_both` and put it in clippy_utils.